### PR TITLE
fix: parse REST API query strings

### LINKED ISSUE #33: [FEATURE] Implement URL query string parsing in REST API server
## Description
The `RestAPIServer::parseQueryString()` method is currently a stub that returns an empty parameters map, preventing REST API endpoints from processing URL query parameters for filtering, pagination, and time range selection.

## Current Behavior
```cpp
std::unordered_map<std::string, std::string>
RestAPIServer::parseQueryString(const std::string& query) {
  std::unordered_map<std::string, std::string> params;
  // Simple query string parsing (not implemented for brevity)
  boost::ignore_unused(query);
  return params;
}
```
Located at: visualization/WebServer.cpp:1000-1006

## Impact

- REST API cannot process query parameters from URLs
- No support for time range filtering (e.g., ?start=123&end=456)
- No pagination support (e.g., ?limit=100&offset=0)
- No metric filtering (e.g., ?metrics=pnl,sharpe_ratio)
- Limits API flexibility and usability

## Expected Behavior

Parse standard URL query strings and return key-value pairs:

Examples:
- Input: "start=1234567890&end=9876543210&limit=100"
- Output: {{"start", "1234567890"}, {"end", "9876543210"}, {"limit", "100"}}
- Input: "symbol=BTC-USD&interval=1h"
- Output: {{"symbol", "BTC-USD"}, {"interval", "1h"}}
- Input: "" (empty or malformed)
- Output: {} (empty map, graceful degradation)

## Implementation Requirements

1. Parse query string format:
  - Split by & delimiter for parameters
  - Split by = for key-value pairs
  - Handle empty/missing values gracefully
2. URL decoding:
  - Decode percent-encoded characters (e.g., %20 → space)
  - Handle special characters properly
  - Support UTF-8 encoding
3. Error handling:
  - Return empty map for null/empty input
  - Skip malformed parameters
  - No exceptions thrown (graceful degradation)
4. Edge cases:
  - Handle parameters without values: key= → {{"key", ""}}
  - Handle duplicate keys (keep last value or create array)
  - Handle ? prefix if present in query string

## Implementation Approach I am Thinking of and Considering Tradeoffs

Option 1: Manual parsing (lightweight)
```cpp
std::unordered_map<std::string, std::string>
RestAPIServer::parseQueryString(const std::string& query) {
  std::unordered_map<std::string, std::string> params;

  if (query.empty()) return params;

  std::istringstream iss(query);
  std::string pair;

  while (std::getline(iss, pair, '&')) {
    auto eqPos = pair.find('=');
    if (eqPos != std::string::npos) {
      std::string key = pair.substr(0, eqPos);
      std::string value = pair.substr(eqPos + 1);
      params[urlDecode(key)] = urlDecode(value);
    }
  }

  return params;
}
```

Option 2: Using Boost

- I can leverage existing Boost dependency
- Use boost::algorithm::split for parsing
- More robust handling

## Additional helper needed:
```cpp
std::string RestAPIServer::urlDecode(const std::string& str);
```
For decoding percent-encoded characters.

Affected API Endpoints

- GET `/api/v1/strategies/{id}/performance?start=X&end=Y`
- GET `/api/v1/strategies/{id}/chart?metric=pnl&range=1h`
- Any future endpoints requiring query parameters

## Testing Considerations

- Test with various query string formats
- Test URL encoding/decoding
- Test empty and malformed inputs
- Test with special characters
- Verify no crashes on edge cases

## Priority

Medium - Needed for full REST API functionality, but endpoints work without it (just with default parameters)

### For easy access of the implementation check

- visualization/WebServer.cpp (implement parseQueryString() and add urlDecode() helper)
- visualization/WebServer.h (add urlDecode() method declaration if needed)

### PR DESCRIPTION
**Pull request by OoreFasawe**
## Summary by CodeRabbit

- **New Features**
  - Improved handling of URL query parameters, including encoded characters, spaces, repeated parameters, and values containing equals signs.
  - Performance requests now work correctly when query strings are included.

- **Bug Fixes**
  - Invalid or incomplete query parameters are safely ignored.
  - Improved URL decoding and routing for REST and static-file requests.
  - Improved container health-check reliability when monitoring the API endpoint.

### CODE REVIEW

**Comment by OoreFasawe** [comment]
## Summary

Implement REST API query-string parsing and support query parameters on routed requests.

## Changes

- Added percent-encoded URL decoding.
- Converted `+` characters to spaces.
- Added support for:
  - Optional leading `?`
  - Multiple `&`-separated parameters
  - Empty and valueless parameters
  - Values containing `=`
  - Duplicate keys with last-value-wins behavior
- Skipped malformed percent-encoded parameters without throwing exceptions.
- Updated request routing to separate paths from query strings.
- Added focused WebServer regression tests.

## Testing

- Built the visualization target successfully.
- Built `web_server_tests` successfully.
- Ran the focused CTest suite successfully.
- Result: 7 tests passed.

## Notes

This change makes query parameters available to REST handlers. Endpoint-specific filtering, pagination, and time-range behavior remain separate follow-up work because the current handlers do not yet apply those parameters.

**Comment by OoreFasawe** [comment]
Hey @chizy7 thanks for the PinnacleMM project! I’ve submitted a PR to parse the REST API query strings. Would appreciate your thoughts whenever you have a chance to review. Let me know if you have any questions!

File: visualization/WebServer.cpp [hunk 1]
```diff
@@ -1018,11 +1070,38 @@ json RestAPIServer::createSuccessResponse(const json& data) {
...
+      continue;
+    }
+
+    params[key] = value;
+  }
+
   return params;
 }
 
+std::string RestAPIServer::urlDecode(const std::string& str) {
```
Comment (chizy7) [inline visualization/WebServer.cpp]: urlDecode is never called anywhere, not in production code and not in the tests. I would remove it for now (we can always bring it back when something actually needs it).

File: visualization/WebServer.cpp [hunk 2]
```diff
@@ -857,33 +906,36 @@ http::response<http::string_body>
...
     // Extract strategy ID
-    auto start = target.find("/api/v1/strategies/") + 19;
-    auto end = target.find("/performance");
-    auto strategyId = target.substr(start, end - start);
-    return handleGetPerformance(strategyId, "");
-  } else if (target == "/api/risk/state") {
+    auto start = path.find("/api/v1/strategies/") + 19;
+    auto end = path.find("/performance");
+    auto strategyId = path.substr(start, end - start);
+    return handleGetPerformance(strategyId, query);
```
Comment (chizy7) [inline visualization/WebServer.cpp]: handleGetPerformance still opens with boost::ignore_unused(query), and parseQueryString is never called outside the tests. So start and end route correctly now but the values do nothing yet. Are you planning to wire them into the handler, or should we track that in a follow-up issue?

File: visualization/WebServer.h [hunk 3]
```diff
@@ -244,12 +245,14 @@ class RestAPIServer : public std::enable_shared_from_this<RestAPIServer> {
...
   json createErrorResponse(const std::string& error, int code = 400);
   json createSuccessResponse(const json& data);
+  std::string urlDecode(const std::string& str);
   std::unordered_map<std::string, std::string>
   parseQueryString(const std::string& query);
   std::string extractPath(const std::string& target);
   std::string getContentType(const std::string& path);
 
   friend class HttpSession;
+  friend class WebServerTests;
```
Comment (chizy7) [inline visualization/WebServer.h]: Small one, not blocking: this friend declaration puts test knowledge into a production header. Since decodeQueryComponent is already a free function, parseQueryString could be one too, and then the friendship is not needed.

**Review by chizy7 — COMMENTED** [review]
Thanks for this. I checked out the branch, built it, ran the test suite and hit the running server with curl. Query strings on the API endpoints used to fall through to the static file handler and 404, and now they route correctly, so the fix is real. A few small comments.

File: visualization/WebServer.cpp (hunk 2 above)
Comment (OoreFasawe) [reply visualization/WebServer.cpp]: I think this should be tracked in a follow-up case, and we can keep this PR small and focused only on parsing.

File: visualization/WebServer.cpp (hunk 1 above)
Comment (OoreFasawe) [reply visualization/WebServer.cpp]: Done.

File: visualization/WebServer.h (hunk 3 above)
Comment (OoreFasawe) [reply visualization/WebServer.h]: Good catch. Added a commit to prevent leaking test information to production header.
